storage/innobase/btr/btr0btr.cc

```diff
@@ -1269,51 +1270,62 @@ btr_read_autoinc(dict_index_t* index)
 
 /** Read the last used AUTO_INCREMENT value from PAGE_ROOT_AUTO_INC,
 or fall back to MAX(auto_increment_column).
-@param[in]	table	table containing an AUTO_INCREMENT column
-@param[in]	col_no	index of the AUTO_INCREMENT column
-@return	the AUTO_INCREMENT value
-@retval	0 on error or if no AUTO_INCREMENT value was used yet */
-ib_uint64_t
-btr_read_autoinc_with_fallback(const dict_table_t* table, unsigned col_no)
+@param table          table containing an AUTO_INCREMENT column
+@param col_no         index of the AUTO_INCREMENT column
+@param mysql_version  TABLE_SHARE::mysql_version
+@param max            the maximum value of the AUTO_INCREMENT column
+@return the AUTO_INCREMENT value
+@retval 0 on error or if no AUTO_INCREMENT value was used yet */
+uint64_t btr_read_autoinc_with_fallback(const dict_table_t *table,
+                                        unsigned col_no, ulong mysql_version,
+                                        uint64_t max)
 {
-	ut_ad(table->persistent_autoinc);
-	ut_ad(!table->is_temporary());
+  ut_ad(table->persistent_autoinc);
+  ut_ad(!table->is_temporary());
 
-	dict_index_t*	index = dict_table_get_first_index(table);
+  uint64_t autoinc= 0, max_autoinc= 0;
+  dict_index_t *index= dict_table_get_first_index(table);
+  mtr_t mtr;
+  mtr.start();
 
-	if (index == NULL) {
-		return 0;
-	}
+  if (buf_block_t *block=
+      buf_page_get(page_id_t(table->space_id, index->page),
+                   table->space->zip_size(), RW_SX_LATCH, &mtr))
+  {
+    autoinc= page_get_autoinc(block->frame);
 
-	mtr_t		mtr;
-	mtr.start();
-	buf_block_t*	block = buf_page_get(
-		page_id_t(index->table->space_id, index->page),
-		index->table->space->zip_size(),
-		RW_S_LATCH, &mtr);
+    if (autoinc > max || mysql_version == 0 || mysql_version < 100210)
```

Comment:
Nice to have the condition (autoinc > max) as well. Perhaps the case when we have already upgraded the version (before this patch) but the value is still junk. 

I also observed that previously we had the check that if the page is not empty and autoinc is still zero. Although I understand that it was introduced for older version,  it might be good to keep that one as well for safety ? Since we don't expect to hit these conditions, we could also want to add some logs.

    auto valid_autoinc = [&]() {
        /* check for max range */
        /* check for non-empty page zero autoinc */
        /* Log warning in case check fails. */
    }

    if (old_version || !valid_autoinc()) {
        ....
    }

Reply:
You are right, we had better invoke `page_set_autoinc()` also when `page_get_autoinc()` returned 0, no matter what the `mysql_version` is. I will add a test for that.